src/record.rs

```diff
@@ -0,0 +1,296 @@
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::sync::{Arc, Mutex};
+
+use crate::TermLike;
+
+/// A Terminal that records all writes and curor moves to it.
+/// Useful for debugging operations on terminals.
+#[cfg_attr(docsrs, doc(cfg(feature = "in_memory")))]
+#[derive(Debug, Clone)]
+pub struct RecordTerm {
+    state: Arc<Mutex<RecordTermState>>,
+}
+
+impl RecordTerm {
+    pub fn new(rows: u16, cols: u16) -> RecordTerm {
+        assert!(rows > 0, "rows must be > 0");
+        assert!(cols > 0, "cols must be > 0");
+        RecordTerm {
+            state: Arc::new(Mutex::new(RecordTermState::new(rows, cols))),
+        }
+    }
+
+    pub fn reset(&self) {
+        let mut state = self.state.lock().unwrap();
+        *state = RecordTermState::new(state.height, state.width);
+    }
+
+    pub fn contents(&self) -> Vec<Move> {
+        self.state.lock().unwrap().history.clone()
+    }
+
+    pub fn cursor_position(&self) -> (i16, i16) {
+        self.state.lock().unwrap().cursor_position
+    }
+}
+
+impl TermLike for RecordTerm {
+    fn width(&self) -> u16 {
+        self.state.lock().unwrap().width
+    }
+
+    fn height(&self) -> u16 {
+        self.state.lock().unwrap().height
+    }
+
+    fn move_cursor_up(&self, n: usize) -> std::io::Result<()> {
+        let mut lock = self.state.lock().unwrap();
+        lock.history.push(Move::Up(n));
+        lock.cursor_position.0 -= i16::try_from(n).unwrap();
+        Ok(())
+    }
+
+    fn move_cursor_down(&self, n: usize) -> std::io::Result<()> {
+        let mut lock = self.state.lock().unwrap();
+        lock.history.push(Move::Down(n));
+        lock.cursor_position.0 += i16::try_from(n).unwrap();
+        Ok(())
+    }
+
+    fn move_cursor_right(&self, n: usize) -> std::io::Result<()> {
+        let mut lock = self.state.lock().unwrap();
+        lock.history.push(Move::Right(n));
+        lock.cursor_position.1 += i16::try_from(n).unwrap();
+        Ok(())
+    }
+
+    fn move_cursor_left(&self, n: usize) -> std::io::Result<()> {
+        let mut lock = self.state.lock().unwrap();
+        lock.history.push(Move::Left(n));
+        lock.cursor_position.1 -= i16::try_from(n).unwrap();
+        Ok(())
+    }
+
+    fn write_line(&self, s: &str) -> std::io::Result<()> {
+        self.write_str(s)?;
+        let mut lock = self.state.lock().unwrap();
+        lock.history.push(Move::NewLine);
+
+        lock.cursor_position.1 = 0;
+        lock.cursor_position.0 += 1;
+        Ok(())
+    }
+
+    fn write_str(&self, s: &str) -> std::io::Result<()> {
```

Comment:
I would maybe add a comment here that this doesn't handle escape sequences. The difficulty in doing so is what ultimately led me to use vt100 for `InMemoryTerm`.

Reply:
ah... hmm yea. that's an issue

Reply:
It's OK, this is just a test mechanism after all. Don't try to implement escape sequence handling - it is nothing but pain, btw.